api/include/opentelemetry/nostd/variant.h

```diff
@@ -0,0 +1,1279 @@
+
+  static_assert(detail::all<!std::is_void<Ts>::value...>::value,
+                "variant can not have a void type as an alternative.");
+
+public:
+  template <typename Front = detail::type_pack_element_t<0, Ts...>,
+            enable_if_t<std::is_default_constructible<Front>::value, int> = 0>
+  inline constexpr variant() noexcept(std::is_nothrow_default_constructible<Front>::value)
+      : impl_(in_place_index_t<0>{})
+  {}
+
+  variant(const variant &) = default;
+  variant(variant &&)      = default;
+
+  template <typename Arg,
+            typename Decayed                                             = decay_t<Arg>,
+            enable_if_t<!std::is_same<Decayed, variant>::value, int>     = 0,
+            enable_if_t<!detail::is_in_place_index<Decayed>::value, int> = 0,
+            enable_if_t<!detail::is_in_place_type<Decayed>::value, int>  = 0,
+            std::size_t I = detail::best_match<Arg, Ts...>::value,
+            typename T    = detail::type_pack_element_t<I, Ts...>,
+            enable_if_t<std::is_constructible<T, Arg>::value, int> = 0>
+  inline constexpr variant(Arg &&arg) noexcept(std::is_nothrow_constructible<T, Arg>::value)
+      : impl_(in_place_index_t<I>{}, std::forward<Arg>(arg))
+  {}
+
+  template <std::size_t I,
+            typename... Args,
+            typename T = detail::type_pack_element_t<I, Ts...>,
+            enable_if_t<std::is_constructible<T, Args...>::value, int> = 0>
+  inline explicit constexpr variant(in_place_index_t<I>, Args &&... args) noexcept(
+      std::is_nothrow_constructible<T, Args...>::value)
+      : impl_(in_place_index_t<I>{}, std::forward<Args>(args)...)
+  {}
+
+  template <
+      std::size_t I,
+      typename Up,
+      typename... Args,
+      typename T = detail::type_pack_element_t<I, Ts...>,
+      enable_if_t<std::is_constructible<T, std::initializer_list<Up> &, Args...>::value, int> = 0>
+  inline explicit constexpr variant(
+      in_place_index_t<I>,
+      std::initializer_list<Up> il,
+      Args &&... args) noexcept(std::is_nothrow_constructible<T,
+                                                              std::initializer_list<Up> &,
+                                                              Args...>::value)
+      : impl_(in_place_index_t<I>{}, il, std::forward<Args>(args)...)
+  {}
+
+  template <typename T,
+            typename... Args,
+            std::size_t I = detail::find_index_sfinae<T, Ts...>::value,
+            enable_if_t<std::is_constructible<T, Args...>::value, int> = 0>
+  inline explicit constexpr variant(in_place_type_t<T>, Args &&... args) noexcept(
+      std::is_nothrow_constructible<T, Args...>::value)
+      : impl_(in_place_index_t<I>{}, std::forward<Args>(args)...)
+  {}
+
+  template <
+      typename T,
+      typename Up,
+      typename... Args,
+      std::size_t I = detail::find_index_sfinae<T, Ts...>::value,
+      enable_if_t<std::is_constructible<T, std::initializer_list<Up> &, Args...>::value, int> = 0>
```

Comment:
Is there some code example that illustrates how I can use this static initializer to populate something like **std::map<std::string, variant>**? i.e. assuming this map-of-variants type is called something like **EventProperties**

Would I be able to do this?
```
EventProperties props(
    {
        { "result", "Success" },
        { "random", rand() },
        { "secret", 5.6872 },
        { "seq", (uint64_t)i },
        ...
    }
);
```
to hold a value object, that I can subsequently use either "diagnostic context" or "resource" to append on event?

Reason why I'm curious about this, is we can share a similar implementation of a collection that'd only requires C++11 anonymous union + type specifier per field. https://en.cppreference.com/w/cpp/language/union

I'm not against backporting the newer std::union, just looking for some guidance where you believe it's gonna be necessary.

Reply:
Yes, you can initialize that way.

It has the same interface as [std::variant](https://en.cppreference.com/w/cpp/utility/variant) from C++17.

Reply:
#60 is how I see it being used in the api.